### PR TITLE
samples: Add samples for async download files #1470

### LINKED ISSUE #1470: Add samples for async download files
We have [samples](https://github.com/googleapis/python-storage/blob/main/samples/snippets/storage_async_upload.py) on how to use async uploads in python, add similar samples for download as well.

### PR DESCRIPTION
**Pull request by chandra-siri**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-storage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary) - NA

Fixes #1470

### CODE REVIEW

File: samples/snippets/storage_async_download.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+"""Sample that asynchronously downloads multiple files from GCS to application's memory.
+"""
+
+
+# [START storage_async_download]
+# This sample can be run by calling `async.run(async_download_blobs('bucket_name'))`
+async def async_download_blobs(bucket_name):
+    """Downloads a number of files in parallel from the bucket;
+        assuming files with prefix `async_sample_blob_%d` exits in bucket."""
+    import asyncio
```
Comment (cojenco) [inline samples/snippets/storage_async_download.py:30]: Let's add an argument description here, as seen in the samples
```suggestion
    # The ID of your GCS bucket
    # bucket_name = "your-bucket-name"
    import asyncio
```

File: samples/snippets/storage_async_download.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    bucket = storage_client.bucket(bucket_name)
+
+    loop = asyncio.get_running_loop()
+
+    tasks = []
+    count = 3
+    for x in range(count):
+        blob_name = f"async_sample_blob_{x}"
+        blob = bucket.blob(blob_name)
+        print('yo!!', blob_name)
```
Comment (cojenco) [inline samples/snippets/storage_async_download.py:43]: I'm guessing this was for testing? Let's remove it :)

File: samples/snippets/storage_async_download.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        # The first arg, None, tells it to use the default loops executor
+        tasks.append(loop.run_in_executor(None, blob.download_as_bytes))
+
+    # If the method returns a value (such as download_as_string), gather will return the values
+    contents = await asyncio.gather(*tasks)
+    for x in range(count):
+        print(f"Downloaded storage object async_sample_blob_{x} from bucket {bucket_name} as the following bytes object: {contents[x]}")
```
Comment (cojenco) [inline samples/snippets/storage_async_download.py:50]: To keep the test logs clean and avoid overloading, let's make the print statement simple. (This is mainly used to assert that the sample completed successfully.)
```suggestion
        print(f"Downloaded storage object async_sample_blob_{x} from bucket {bucket_name}")
```

File: samples/snippets/storage_async_download.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    tasks = []
+    count = 3
+    for x in range(count):
+        blob_name = f"async_sample_blob_{x}"
+        blob = bucket.blob(blob_name)
+        print('yo!!', blob_name)
+        # The first arg, None, tells it to use the default loops executor
+        tasks.append(loop.run_in_executor(None, blob.download_as_bytes))
+
+    # If the method returns a value (such as download_as_string), gather will return the values
```
Comment (cojenco) [inline samples/snippets/storage_async_download.py:47]: nit: I think it'd make more sense since we're using download_as_bytes here
```suggestion
    # If the method returns a value (such as download_as_bytes), gather will return the values
```

**Review by cojenco — COMMENTED** [review]
Thanks for adding this! The sample LGTM. Left a few comments; Feel free to let me know if you have any questions.

**Comment by cojenco** [comment]
Added the label `kokoro:force-run` to rerun the presubmit tests. 

The tests are passing now that the core library team fixed a key rotation in one of the kokoro projects, fyi internal b/413445927

File: samples/snippets/storage_async_download.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+"""Sample that asynchronously downloads multiple files from GCS to application's memory.
+"""
+
+
+# [START storage_async_download]
+# This sample can be run by calling `async.run(async_download_blobs('bucket_name'))`
+async def async_download_blobs(bucket_name):
+    """Downloads a number of files in parallel from the bucket;
+        assuming files with prefix `async_sample_blob_%d` exits in bucket."""
```
Comment (bajajneha27) [inline samples/snippets/storage_async_download.py:29]: QQ: is this really a requirement for async download?

File: samples/snippets/storage_async_download.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py:1]: I kept it consistent with other samples , see this [samples/snippets/storage_async_upload.py ](https://github.com/googleapis/python-storage/blob/63c1139765caa788c0502d6f2962d8030219f751/samples/snippets/storage_async_upload.py#L1) for example.

File: samples/snippets/storage_async_download.py (hunk 5 above)
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py:29]: Are you talking about this ->  _assuming files with prefix `async_sample_blob_%d` exits in bucket_ ?

If yes, I made that assumption because I wanted to keep the code in sample as simple & minimal as possible. Otherwise I've to add either of these

1. create objects 
2. ask user to pass the list of objects as input

Both of them will add complexity. 2nd one is a bit less complex but again we need to add unit tests for couple of edge cases in addition to testing the core functionality. 

LMK what do you think, please resolve comment if it sounds good to you.

File: samples/snippets/storage_async_download.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+import asyncio
+import sys
+
+
+"""Sample that asynchronously downloads multiple files from GCS to application's memory.
+"""
+
+
+# [START storage_async_download]
```
[the remark this answers is not in the document]
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py:25]: I did a quick search across googleapis , looks like we don't have anything related to async upload or download

https://github.com/search?q=org%3Agoogleapis+%2F%5C%5BSTART+storage_.*async%2F&type=code

File: samples/snippets/storage_async_download.py (hunk 2 above)
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py:45]: Sorry, my bad.

File: samples/snippets/storage_async_download.py (hunk 4 above)
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py:49]: changed it to `download_as_bytes` in the comments.

File: samples/snippets/storage_async_download.py (hunk 3 above)
Comment (cojenco) [reply samples/snippets/storage_async_download.py]: Reopening; Could you remove `contents[x]` and replace with a simple print statement, thanks!
`print(f"Downloaded storage object async_sample_blob_{x} from bucket {bucket_name}")`

The main concern here is we don't want to print contents for best practice. There has been incidents where testing logs get overloaded because tests accidentally printed contents.

**Review by cojenco — APPROVED** [review]
Please address the unresolved comments before merging; otherwise LGTM, thanks!

File: samples/snippets/storage_async_download.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cojenco) [reply samples/snippets/storage_async_download.py:1]: In python, the samples can run via the command-line, so a shebang is included as the first line, followed by the copyright. Reference:https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#shebang

File: samples/snippets/storage_async_download.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python
+
+# Copyright 2025 Google Inc. All Rights Reserved.
```
Comment (cojenco) [inline samples/snippets/storage_async_download.py:3]: Sorry I missed this earlier. This is probably remnants from previous older samples. Please help update to 
go/releasing/preparing#Apache-header linked in go/copyright

The current copyright starts with
`Copyright 2025 Google LLC`

Comment (chandra-siri) [reply samples/snippets/storage_async_download.py:3]: As I understand there's no action required from my side, right?  ( Ref go/copyright#build-shell-scripts-sawzall )

Comment (bajajneha27) [reply samples/snippets/storage_async_download.py:3]: @chandra-siri What you have is `Copyright 2025 Google Inc` , but the current copyright as Cathy mentioned is `Copyright 2025 Google LLC`

File: samples/snippets/storage_async_download.py (hunk 5 above)
Comment (bajajneha27) [reply samples/snippets/storage_async_download.py:29]: I just mean that is it necessary to have the prefix? or any object would do?

**Review by bajajneha27 — APPROVED** [review]
LGTM. Please resolve the pending comments.

File: samples/snippets/storage_async_download.py (hunk 5 above)
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py:29]: Prefix is necessary.

File: samples/snippets/storage_async_download.py (hunk 3 above)
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py]: Done.

File: samples/snippets/storage_async_download.py (hunk 8 above)
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py]: Done.

File: samples/snippets/storage_async_download.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    tasks = []
+    for file_name in file_names:
+        blob = bucket.blob(file_name)
+        # The first arg, None, tells it to use the default loops executor
+        tasks.append(loop.run_in_executor(None, blob.download_as_bytes))
+
+    # If the method returns a value (such as download_as_bytes), gather will return the values
+    _ = await asyncio.gather(*tasks)
+    for i, file_name in enumerate(file_names):
```
Comment (bajajneha27) [inline samples/snippets/storage_async_download.py:51]: nit: `i` is not required here.

**Review by bajajneha27 — APPROVED** [review]

File: samples/snippets/storage_async_download.py (hunk 5 above)
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py]: Added 1 more param - *file_names

File: samples/snippets/storage_async_download.py (hunk 9 above)
Comment (chandra-siri) [reply samples/snippets/storage_async_download.py]: Done.
